<!DOCTYPE HTML>
<html>
	<head>
		<meta charset="utf-8">
		<title>GeneTargeter</title>
		<!--link rel="apple-touch-icon" sizes="57x57" href="static/assets/favicon/apple-icon-57x57.png">
		<link rel="apple-touch-icon" sizes="60x60" href="static/assets/favicon/apple-icon-60x60.png">
		<link rel="apple-touch-icon" sizes="72x72" href="static/assets/favicon/apple-icon-72x72.png">
		<link rel="apple-touch-icon" sizes="76x76" href="static/assets/favicon/apple-icon-76x76.png">
		<link rel="apple-touch-icon" sizes="114x114" href="static/assets/favicon/apple-icon-114x114.png">
		<link rel="apple-touch-icon" sizes="120x120" href="static/assets/favicon/apple-icon-120x120.png">
		<link rel="apple-touch-icon" sizes="144x144" href="static/assets/favicon/apple-icon-144x144.png">
		<link rel="apple-touch-icon" sizes="152x152" href="static/assets/favicon/apple-icon-152x152.png">
		<link rel="apple-touch-icon" sizes="180x180" href="static/assets/favicon/apple-icon-180x180.png">
		<link rel="icon" type="image/png" sizes="192x192"  href="static/assets/favicon/android-icon-192x192.png">
		<link rel="icon" type="image/png" sizes="32x32" href="static/assets/favicon/favicon-32x32.png">
		<link rel="icon" type="image/png" sizes="96x96" href="static/assets/favicon/favicon-96x96.png">
		<link rel="icon" type="image/png" sizes="16x16" href="static/assets/favicon/favicon-16x16.png">
		<link rel="manifest" href="static/assets/favicon/manifest.json">
		<meta name="msapplication-TileColor" content="#ffffff">
		<meta name="msapplication-TileImage" content="static/favicon/ms-icon-144x144.png">
		<meta name="theme-color" content="#ffffff"-->
		<link rel="shortcut icon" href="{{ url_for('static', filename='assets/favicon.ico') }}">

		<link rel="stylesheet" href="static/css/style_v1.css">
		<script type="text/javascript" src="static/js/jszip.js"></script>
		<script type="text/javascript" src="static/js/FileSaver.min.js"></script>
		<script type="text/javascript" src="static/js/main_v1.js"></script>
		<script type="text/javascript" src="static/js/EPPZScroll.js"></script>
	</head>

	<body onload="init()" onscroll="scrollEvt()">
		<label class="roundBtn hidden" id="backToTopBtn" onclick="EPPZScrollTo.scrollVerticalToElementById('Title', 0)">
			<span id="upBtn">^</span></label>

		<div class="headTitle" id="Title" onclick="EPPZScrollTo.scrollVerticalToElementById('controls', 50)">
			<span id="titLeft">Gene</span><span id="titRight">Targeter</span>
		</div>

		<div class="mainText">
			<p>
          Creates custom gene-editing constructs
          designed to deliver DNA payload to specific genes in malaria parasites,
					including for knock-out or conditional knock-down in <a href="http://www.who.int/mediacentre/factsheets/fs094/en/" target="_blank"><i>Plasmodium falciparum</i></a>.
      </p>
			<p>
					Preprint available on <a href="https://doi.org/10.21203/rs.3.rs-565539/v1" target="_blank">ResearchSquare</a>.
					Command line interface and code available through <a href="https://github.com/pablocarderam/GeneTargeter" target="_blank">
						GitHub.</a> Developed by the
	          <a href="http://web.mit.edu/nileslab/" target="_blank">Niles Lab at MIT</a>.
			</p>
		</div>

		<div id="App" class="app" onload="init()">
			<!-- <script src="//cdnjs.cloudflare.com/ajax/libs/socket.io/2.0.4/socket.io.js"></script> -->
			<script type="text/javascript" src="static/js/socket.io.js"></script>
			<script src="static/js/socketHandler_v0.js"></script>
			<!--script src="static/js/app.js"></script-->
			<img src="static/assets/instructions.png" width="50%">
				<br>
			<div class="small" id="controls">
				<p>
					Choose a pre-loaded targeting system: &nbsp<select id="plasmidType" onchange="changeUTRTarget()">
						<option value="pSN054_V5" selected="selected">pSN054_V5 (3' knockdown)</option>
						<option value="pSN054">pSN054 (3' knockdown)</option>
						<option value="pSN150">pSN150 (5' knockdown)</option>
						<option value="pSN150-KO">pSN150 (knockout)</option>
						<option value="custom">Custom plasmid</option>
						<!-- <option value="pSN150-Ter">pSN150 with Nhe1-free T7 terminator (5' payload)</option> -->
						<!-- <option value="pSN150-KO-Ter">pSN150 with Nhe1-free T7 terminator (KO construct)</option> -->
					</select>
				</p>
				<p> <b><i>or</i></b> </p>
				<p>
					<label class="btn" ><input type="file" id="plasmidFileForm" accept=".gb" multiple size="50" onchange="uploadPlasmidFile()">
							<span> <a class="tooltip" target="_blank"
								tipTxt='Custom plasmids must contain annotations labeled "LHR", "RHR", "sgRNA
								Sequence", and "sgRNA Cassette" (used to define a synthesis fragment).
								An optional "Recoded Region" annotation can be included as well. These
								regions will be replaced with the sequences specific to the target gene.'>
									Upload a custom base plasmid .gb file
								</a></span>
						</label><br>
						<span> <a class="tooltip" target="_blank" href="static/assets/custom_genetargeter_example_psn054_v5.gb"
							tipTxt='Custom plasmids must contain annotations labeled "LHR", "RHR", "sgRNA
							Sequence", and "sgRNA Cassette" (used to define a synthesis fragment).
							An optional "Recoded Region" annotation can be included as well. These
							regions will be replaced with the sequences specific to the target gene.'>
								(Download a custom base plasmid example here)
							</a></span>
				</p>
				<p>
					<div id="selectedPlasmidFile"></div>
				</p>
				<p>
					Target location within gene: &nbsp<select id="locationType" onchange="changeUTRTarget()">
						<option value="5prime">5' extreme</option>
						<option value="3prime" selected="selected">3' extreme</option>
						<option value="center">Center (no recoded region)</option>
					</select>
				</p>
				<p>
					<label class="btn" id="optionsBtn" onclick="toggleOptions()">Show more options</label>
				</p>
				<div id=moreOptions class="expand">
					<div class="measuringWrapper">
						<br>
						<p>
							Guide RNAs will be selected automatically based
							on <a href="https://dx.doi.org/10.1038/nbt.3437" target="_blank">
							Doench et al. (2016)</a> on-target score and either
							<a href="https://dx.doi.org/10.1038/nbt.3437" target="_blank">
							Doench et al. (2016)</a> or
							<a href="http://crispr.mit.edu/about" target="_blank">
							Hsu et al. (2013)</a> off-target scores. Off-target scores
							are evaluated based on off-target gRNAs in the <i>P. falciparum</i>
							3D7 genome.
						</p>
						<p>
							Alternatively, you may also
							manually annotate a target region to be completely removed. The
							target region must be labeled as "Target Region" on the input gene
							GenBank file. All sgRNAs will be selected from this region, and
							the LHR and RHR will be selected outside of it. If the base
							plasmid system includes a 5' or 3' recoded region, it will
							comprise at least the entirety of the annotated target region.
						</p>
						<p>
							Finally, you may
							manually annotate all
							<a href="https://benchling.com" class="tooltip" target="_blank"
								tipTxt="You can use Benchling's CRISPR tool, but you must
								create annotations on the sequence for all the gRNAs to be
								considered by GeneTargeter.">
								CRISPR guide RNAs</a>
							to be included in targeting. Guide RNA annotations should start
							with the word "gRNA" in the label, followed by
							<a href="https://benchling.com" class="tooltip" target="_blank"
								tipTxt="i.e., your best-ranked gRNA should be labeled
								'MyGene gRNA 1', your second best should be 'MyGene gRNA
								2, etc.'" class="tooltip">
								a number</a>
							indicating their order of preference. You may also annotate
							hand-picked left and right homologous regions on your file
							(annotated with the words "LHR" and "RHR" in the label,
							respectively), if
							you do not want GeneTargeter to choose them automatically for you.
						</p><br>

						<p>
							<p class="head">Manual HR/sgRNAs</p>
							Use user-annotated target region, homology regions and/or guide RNAs (if any):
							<input class="chkBox" type="checkbox" unchecked id="HRannChkBox">
						</p>

						<p class="head">Bulk processing options:</p><p>
							<a href="" class="tooltip" target="_blank"
								tipTxt="Useful for bulk processing: aggregates all gene outputs into a single file for each output type, containing all gene outputs of that given file type; oligos are sorted by purpose and orientation before gene.">
								Consolidate files by type:</a>
							<input class="chkBox" type="checkbox" unchecked id="bulkFileChkBox">
							&nbsp <a href="" class="tooltip" target="_blank"
								tipTxt="If not *None*, adds this prefix to all file, oligo, DNA fragment, and GenBank names.">
								Project prefix:</a> <input class="txtInput" type="text" id="prefix" value="*None*" style="width:50px">
							&nbsp <a href="" class="tooltip" target="_blank"
								tipTxt="If prefix is not *None* and a number is given here, numbers each gene processed consecutively starting at the given value; adds number to prefix on all file, oligo, DNA fragment, and GenBank names.">
								Start numbering on:</a> <input class="txtInput" type="number" id="prefixNum" value="">
						</p>

						<p>
							Treat genes as coding/noncoding: &nbsp<select id="setCoding">
								<option value="Coding">Coding</option>
								<option value="Noncoding">Noncoding</option>
								<option value="Auto" selected="selected">Auto</option>
							</select>
						</p><br>

						<div class="container">
							<div class="column-center">
								<p class="head">Right chromosomal integration homologous region</p>
								Size:<br>
								&nbsp&nbsp Minimum: &nbsp&nbsp<input class="txtInput" type="number" id="RHRMin" value="400"> bp <br>
								&nbsp&nbsp Preferred: <input class="txtInput" type="number" id="RHRPref" value="500"> bp <br>
								&nbsp&nbsp Maximum: &nbsp&nbsp<input class="txtInput" type="number" id="RHRMax" value="750"> bp <br>
								<p>
									<span id="maxDistRHRTxt">Maximum distance from gene: </span>
									<input class="txtInput" type="number" id="maxDistRHR" value="125">
									bp
								</p>
								<p>
									Search for best start/end point
									<input class="txtInput" type="number" id="optLowRHR" value="20">
									bp upstream and
									<input class="txtInput" type="number" id="optHighRHR" value="20">
									bp downstream of first valid start/end point
								</p>
								<p>
									Minimum melting temperature of the first and last
									<input class="txtInput" type="number" id="endsRHR" value="40"> bp
									of the right homologous region:
									<input class="txtInput" type="number" id="endsTempRHR" value="55"> ºC
								</p>
							</div>
							<div class="column-left">
								<p class="head">Left chromosomal integration homologous region</p>
								Size:<br>
								&nbsp&nbsp Minimum: &nbsp&nbsp<input class="txtInput" type="number" id="LHRMin" value="400"> bp <br>
								&nbsp&nbsp Preferred: <input class="txtInput" type="number" id="LHRPref" value="500"> bp <br>
								&nbsp&nbsp Maximum: &nbsp&nbsp<input class="txtInput" type="number" id="LHRMax" value="750"> bp <br>
								<p>
									<span id="maxDistLHRTxt">Maximum distance from gRNA: </span>
									<input class="txtInput" type="number" id="maxDistLHR" value="700"> bp
								</p>
								<p>
									Search for best start/end point
									<input class="txtInput" type="number" id="optLowLHR" value="20">
									bp upstream and
									<input class="txtInput" type="number" id="optHighLHR" value="20">
									bp downstream of first valid start/end point
								</p>
								<p>
									Minimum melting temperature of the first and last
									<input class="txtInput" type="number" id="endsLHR" value="40"> bp of
									the left homologous region:
									<input class="txtInput" type="number" id="endsTempLHR" value="55"> ºC
								</p>
							</div>
							<div class="column-right">
								<p class="head">Gibson Assembly</p>
								Annealing region size:
								&nbsp&nbsp Minimum: &nbsp&nbsp<input class="txtInput" type="number" id="gibMin" value="20"> bp <br>
								&nbsp&nbsp Preferred: <input class="txtInput" type="number" id="gibPref" value="30"> bp <br>
								&nbsp&nbsp Maximum: &nbsp&nbsp<input class="txtInput" type="number" id="gibMax" value="60"> bp <br>
								<br>
								*Gibson homology region is taken as above preferred size.
								<p>
									Minimum primer melting temperature:
									<input class="txtInput" type="number" id="gibTemp" value="50"> ºC
								</p>
								<p>
									Maximum melting temperature difference between primers:
									<input class="txtInput" type="number" id="gibTDif" value="5"> ºC
								</p>
							</div>
						</div>
						<div class="container">
							<div class="column-leftcenter">
								<p class="head">Guide RNA (gRNA) Selection</p>
								<p>
									Minimum GC content: &nbsp<input class="txtInput" type="number" id="gRNAGCContent" value="25"> % <br>
									CRISPR cutting enzyme:
											<select id="enzymeType" onchange="changeEnzyme()">
												<option value="Cas9" selected="selected">Cas9 (3' PAM)</option>
												<option value="Cas12">Cas12 (aka Cpf1; 5' PAM)</option>
											</select> <br>
									PAM Sequence: &nbsp<select id="PAMSequence">
										<option value="NGG" selected="selected">NGG (SpCas9)</option>
										<option value="NAG">NAG (SpCas9, low efficiency)</option>
										<option value="NAAG">NAAG (SpCas9 QQR1)</option>
										<option value="TGAG">TGAG (SpCas9 VQR/EQR)</option>
										<option value="NGCG">NGCG (SpCas9 VRER)</option>
										<option value="NNNNGATT">NNNNGATT (NmCas9)</option>
										<option value="NNAGAAW">NNAGAAW (St1Cas9)</option>
										<option value="NAAAAC">NAAAAC (TdCas9)</option>
										<option value="NNGRRT">NNGRRT (SaCas9)</option>
										<option value="NNGRR">NNGRR (SaCas9, low efficiency)</option>
										<option value="TTTV">TTTV (AsCas12)</option>
										<option value="TYCV">TYCV (AsCas12 S542R)</option>
										<option value="CCCC">CCCC (AsCas12 K607R)</option>
										<option value="TATV">TATV (AsCas12 S542R/K548V/N552R)</option>
									</select> <br>
									On-target <a href="#bottom" class="tooltip"
										tipTxt="Doench et al. (2016) scores obtained from gradient-boosted regression trees of SpCas9 data. Kim et al. (2017) scores determined through logistic regression of AsCas12 data.">
										scoring method</a>:
											<select id="gRNAOnTargetMethod" onchange="changeOnScoringMethod()">
												<option value="azimuth" selected="selected">Azimuth (Doench et al., 2016)</option>
												<!-- <option value="ruleset2">Rule Set 2 (Doench et al., 2016)</option> -->
												<option value="cindel">CINDEL (Kim et al., 2017)</option>
											</select> <br>
									Minimum on-target score: &nbsp<input class="txtInput" type="number" id="gRNAOnTargetCutoff" value="35"> % <br>
									Off-target <a href="http://crispr.mit.edu/about" class="tooltip" target="_blank"
										tipTxt="CFD scores are more accurate in Cas9-edited mammalian cells (Doench et al., 2016). Hsu et al. (2013) scores are used by crispr.mit.edu and Benchling, and are used as proxies for Cas12. Individual scores are aggregated as per crispr.mit.edu.">
										scoring method</a> for single hits:<br>
											<select id="gRNAOffTargetMethod" onchange="changeOffScoringMethod()">
												<option value="cfd" selected="selected">CFD (Doench et al., 2016)</option>
												<option value="hsu">Zhang Lab (Hsu et al. 2013)</option>
											</select> <br>
									Minimum off-target total score: </a>&nbsp<input class="txtInput" type="number" id="minOffTargetScore" value="20"> %<br>
									Maximum single hit off-target score:</a>&nbsp<input class="txtInput" type="number" id="maxOffTargetHitScore" value="50"> %<br>
								</p>
								<br><br>
							</div>
							<div class="column-right">
								<p class="head">
									Recoded Region
								</p>
								<p>
									Codon optimize to
									<select id="codonOptimizeOrg">
										<option value="T. gondii" selected="selected">T. gondii</option>
										<option value="P. falciparum 3D7">P. falciparum 3D7</option>
										<option value="P. vivax">P. vivax</option>
										<option value="E. coli K12">E. coli K12</option>
										<option value="S. cerevisiae">S. cerevisiae</option>
										<option value="H. sapiens">H. sapiens</option>
										<option value="R. norvegicus">R. norvegicus</option>
										<option value="scramble">Scramble</option>
									</select>
									<a href="http://omicsomics.blogspot.com/2009/04/is-codon-optimization-bunk.html" target="_blank">
										using</a>
									<select id="codonOptimStrat">
										<option value="CAI Maximization" selected="selected">CAI Maximization</option>
										<option value="Codon Sampling">Codon Sampling</option>
									</select>
								</p>
								<p>
									Minimum gBlock size:
									<input class="txtInput" type="number" id="minFragSize" value="250">
									bp
								</p>
								<p>
									Maximum gBlock size:
									<input class="txtInput" type="number" id="maxFragSize" value="1000">
									bp
								</p>
								<p>
									Extend gBlock if under minimum size (instead of using oligos):
									<input type="checkbox" id="gBlockDefault" checked>
								</p>
								<p>
									Use HA tags (5' designs): <select id="haTag">
										<option value="Yes">Yes</option>
										<option value="No">No</option>
										<option value="Auto" selected="selected">Auto</option>
									</select>
								</p>
							</div>
							<br><br><br>
						</div>
					</div>
				</div>
			</div>

			<br>

			<p>
				Provide a list of 3D7 gene IDs or upload a
				<a href="https://benchling.com" class="tooltip" target="_blank"
					tipTxt="A GenBank file (.gb) can
					be downloaded from Benchling clicking on the Information
					icon on the righthand tool menu and scrolling to the
					'Download' section.">
					GenBank file</a>
				annotated with the gene's sequence
				<a href="https://benchling.com" class="tooltip" target="_blank"
					tipTxt="Benchling displays 5'->3'
					annotations from left to right on the top strand. If your
					gene appears from right to left on the bottom strand, you
					can select the whole sequence -> right click -> Copy Special
					-> Reverse Complement -> paste in new file.">
					in 5'->3' sense.</a>
					If you load your own files, make sure the target gene name in each
					text box matches the gene's annotation on the plasmid.
				</p>

			<p>
				<label class="btn" onclick="geneIDSubmissionInterface()">
					<span> Select genes </span></label>
					or
					<label class="btn" ><input type="file" id="geneFileForm" accept=".gb" multiple size="50" onchange="uploadGeneFile()">
					<span> Upload .gb files </span></label>
			</p>

			<p>
				<label class="runBtn" id="run" onclick="askCredentials()">Select one or more GenBank gene files.</label>
			</p>

			<div class="fileList small" id="selectedFiles">
			</div>

			<p>
				<div class="fileList" id="outputLog"></div>
			</p>

			<p>
				GeneTargeter outputs seven kinds of files:
				<ul>
					<li>
						<b>a GenBank file</b> containing the given gene annotated with
						the chosen left and right homologous regions, before chromosomal
						editing.
					</li>
					<li>
						<b>a GenBank file</b> containing the fully annotated plasmid
						vector designed to target the given gene.
					</li>
					<li>
						<b>a GenBank file</b> containing the edited chromosomal locus
						targeted by the construct.
					</li>
					<li><b>a CSV file</b> containing primers and oligos designed to
						assemble the new plasmid using
						<a href="https://dx.doi.org/10.1038/nmeth.1318" target="_blank">
							Gibson Assembly</a>.
						This file can be
						<a href="https://benchling.com" class="tooltip" target="_blank"
							tipTxt="Open the '+' drop-down menu in the top-right corner of
							Benchling's interface, select 'More', and select 'Import oligos'.
							Copy and paste the CSV file contents onto Benchling's field.">
							imported into Benchling</a>. Abbreviates primer names to fit on
							commercial tube labels with the <a href="" class="tooltip" tipTxt="The seven digit gene identifier
							code follows 'PF3D7_'.
							Oligo types include:
							LHR (Gibson overhang PCR primers)
							RHR (Gibson overhang PCR primers)
							gRNA (Klenow oligos for gRNA sequence)
							gBlock (gBlock sequencing primer)
							RecKlen (Klenow oligos for recoded region, if the region is small
							enough)
							Orientation refers to forward (F) and reverse (R) primers.">format:</a><br>

							Seven Digit Gene Identifier_Oligo type_Orientation
					</li>
					<li><b>a FASTA file</b> containing gene fragments with overlaps
						designed to assemble the new plasmid using
						<a href="https://dx.doi.org/10.1038/nmeth.1318" target="_blank">
							Gibson Assembly</a>.
					</li>
					<li>
						<b>a CSV file</b> comparing different possible sgRNAs evaluated
						by GeneTargeter before making a choice according to a variety of
						scoring metrics, as well as the corresponding
						<a class="tooltip"
						tipTxt="sgRNAs are either excluded or recoded through codon
						switching in the final design to avoid CRISPR enzyme recutting.">
						recoded sgRNA</a> for each in the final design.
					</li>
					<li>
						<b>a text file</b> containing a message log and warnings
						issued during the <i>in silico</i> design process, along with
						plasmid assembly instructions.
					</li>
				</ul>
			</p>

			<div id="myModal" class="modal">

			  <!-- Modal content -->
			  <div class="modal-content" id="modContent">
			    <div class="modal-header">
			      <span class="close" onclick="closeModal('modContent','myModal')">×</span>
			      <h2>Hold up though...</h2>
			    </div>
			    <div class="modal-body">
						<p>
							Don't forget to allow your browser to download files! To minimize
							the number of pop-up windows, you can check your browser's
							settings and preferences before you run to make sure you allow
							your browser to save all plain text document downloads.
						</p>
						<!-- <h3>Also...</h3>
						<p>
						  GeneTargeter uses DNA sequences that are currently not
						  available to the public. The full software will be
						  publicly available in the near future.
						</p>
						<p>
						  If you have the credentials to download the output, input
						  your passcode here:
						</p>
						<p>
							<input type="text" id="passcode" value="" onkeydown="checkKey()">
						</p> -->
						<p>
							<label class="runBtn" id="submitPassword" onclick="verifyCredentials()">Submit</label>
						</p>
			    </div>
			    <div class="modal-footer" id="modFooter">
			      <h3>.</h3>
			    </div>
			  </div>

			</div>

			<div id="errorWindow" class="modal">
				<div class="modal-content" id="errorWindowContent">
					<div class="modal-header">
						<h2>Server error</h2>
					</div>
					<div class="modal-body">
						<p id="errorContent"></p>
						<p>
							Reload the page to submit a query again.
						</p>
						<div class="modal-footer" id="modFooterError"></div>
					</div>
				</div>
			</div>

			<div id="geneIDSubmission" class="modal">
				<div class="modal-content" id="geneIDSubmissionContent">
					<div class="modal-header">
						<span class="close" onclick="closeModal('geneIDSubmissionContent','geneIDSubmission')">×</span>
						<h2>Submit gene IDs</h2>
					</div>
					<div class="modal-body">
						<p>
							List the gene IDs you want to target below (one ID per line):
						</p>
							<textarea id="geneIDSubmissionTxt" rows="10" cols="30" wrap="soft"></textarea>
							<br>
							<p id="geneIDSubmissionError"></p>
							<br>
						<div class="modal-footer" id="modFooterSelect">
							<label class="btn" onclick="submitGeneIDs()">
								<span> Select genes </span></label>
							<!-- If you want to submit a modified gene file or if your gene isn't pre-loaded here,
							<label class="btn" ><input type="file" accept=".gb" multiple size="50" onchange="uploadFile()">
							<span> upload your own .gb files</span></label>. -->
						</div>
					</div>
				</div>
			</div>

		</div>

		<div class="bottom" id="bottom">
			<a href="http://web.mit.edu/nileslab/" target="_blank">
				<p>
					<img src="static/assets/Logo_NilesLab_GC.png" width="400px">
				</p>
				<!-- <p>
					<img src="static/assets/WhiteLogoNilesLab.png" width="400px">
				</p> -->
			</a>
			<p>
					On-target Cas9 gRNA scoring code and CFD scoring matrix obtained December 27, 2016 from <a href="https://dx.doi.org/10.1038/nbt.3437" target="_blank">Doench et al. (2016)</a> supplementary material.
			</p>
			<p>
					On-target Cas12 gRNA scoring coefficients obtained from <a href="https://dx.doi.org/10.1038/nmeth.4104" target="_blank">Kim et al. (2017)</a> supplementary material, equivalent to the <a href="http://big.hanyang.ac.kr/cindel/" target="_blank">CINDEL</a> online tool. Exact on-target scores may vary between GeneTargeter and <a href="http://big.hanyang.ac.kr/cindel/" target="_blank">CINDEL</a> due to minor differenes in gRNA free energy calculations.
			</p>
			<p>
					Self-folding Gibbs free energy of gRNAs (used to calculate <a href="https://dx.doi.org/10.1038/nmeth.4104" target="_blank">Kim et al. (2017)</a> <a href="http://big.hanyang.ac.kr/cindel/" target="_blank">CINDEL</a> scores) calculated using RNAfold obtained from <a href="http://www.tbi.univie.ac.at/RNA/index.html" target="_blank">ViennaRNA v2.3.3</a> <a href="https://dx.doi.org/10.1186/1748-7188-6-26" target="_blank">(Lorenz et al., 2011)</a>.
			</p>
			<p>
					Off-target Zhang Lab scoring matrix obtained December 27, 2016 from the online <a href="http://crispr.mit.edu/about">MIT CRISPR tool</a> <a href="https://dx.doi.org/10.1038/nbt.2647" target="_blank">(Hsu et al., 2016)</a>.
			</p>
			<p>
				Off-target gRNA database built from <i>P. falciparum</i> genome first published by
				<a href="https://dx.doi.org/10.1038/nature01097" target="_blank"> Gardner et al. (2002)</a>
				and subsequently edited by the <a href="http://plasmodb.org/plasmo/" target="_blank">PlasmoDB</a> community.
			</p>
			<p>
					Codon frequency tables obtained from the <a href="https://hive.biochemistry.gwu.edu/cuts/about" target="_blank">High-performance Integrated Virtual Environment-Codon Usage Tables (HIVE-CUT) database</a> <a href="https://dx.doi.org/10.1186/s12859-017-1793-7" target="_blank">(Athey et al., 2017)</a>.
			</p>
			<p>
					Niles lab logo design by Gaël Chambonnier.
			</p>
			<p>
	        <!-- Background Image: Charisse Flerida Pasaje -->
	        Background Image: <a href="https://commons.wikimedia.org/wiki/File:Plasmodium_falciparum_in_a_patient_presenting_with_pyrexia_and_headache.jpg" target="_blank">
	            <i>Plasmodium falciparum in a patient presenting with pyrexia and
	            headache.</i> O. Erhabor (2013)</a>
	        <br>
	        <a href="https://creativecommons.org/licenses/by-sa/3.0/deed.en" target="_blank">
	            CC BY-SA 3.0</a>
	    </p>
			<br><br>
	    <p>
	        All other code: <a href="http://opensource.org/licenses/MIT" target="_blank">
					MIT License</a>, appropriately.
				<br>
				<a href="https://github.com/pablocarderam/GeneTargeter" target="_blank">
					GitHub src here.</a>
	            <br>
	            Developed by Pablo Cárdenas Ramírez
	            <br>
	            <a href="mailto:pablocarderam@gmail.com?Subject=Hi%20there" target="_top" class="pcr">
					pablocarderam@gmail.com</a>
			</p>
		</div>
	</body>
</html>
