<!DOCTYPE HTML>
<html>
	<head>
		<meta charset="utf-8">
		<title>GeneTargeter</title>
		<!--link rel="apple-touch-icon" sizes="57x57" href="static/assets/favicon/apple-icon-57x57.png">
		<link rel="apple-touch-icon" sizes="60x60" href="static/assets/favicon/apple-icon-60x60.png">
		<link rel="apple-touch-icon" sizes="72x72" href="static/assets/favicon/apple-icon-72x72.png">
		<link rel="apple-touch-icon" sizes="76x76" href="static/assets/favicon/apple-icon-76x76.png">
		<link rel="apple-touch-icon" sizes="114x114" href="static/assets/favicon/apple-icon-114x114.png">
		<link rel="apple-touch-icon" sizes="120x120" href="static/assets/favicon/apple-icon-120x120.png">
		<link rel="apple-touch-icon" sizes="144x144" href="static/assets/favicon/apple-icon-144x144.png">
		<link rel="apple-touch-icon" sizes="152x152" href="static/assets/favicon/apple-icon-152x152.png">
		<link rel="apple-touch-icon" sizes="180x180" href="static/assets/favicon/apple-icon-180x180.png">
		<link rel="icon" type="image/png" sizes="192x192"  href="static/assets/favicon/android-icon-192x192.png">
		<link rel="icon" type="image/png" sizes="32x32" href="static/assets/favicon/favicon-32x32.png">
		<link rel="icon" type="image/png" sizes="96x96" href="static/assets/favicon/favicon-96x96.png">
		<link rel="icon" type="image/png" sizes="16x16" href="static/assets/favicon/favicon-16x16.png">
		<link rel="manifest" href="static/assets/favicon/manifest.json">
		<meta name="msapplication-TileColor" content="#ffffff">
		<meta name="msapplication-TileImage" content="static/favicon/ms-icon-144x144.png">
		<meta name="theme-color" content="#ffffff"-->
		<link rel="shortcut icon" href="{{ url_for('static', filename='assets/favicon.ico') }}">

		<link rel="stylesheet" href="static/css/style_v2.css">
		<script src="static/js/main_v2.js"></script>
		<script src="static/js/EPPZScroll.js"></script>
	</head>

	<body onload="init()" onscroll="scrollEvt()">
		<label class="roundBtn hidden" id="backToTopBtn" onclick="EPPZScrollTo.scrollVerticalToElementById('Title', 0)">
			<span id="upBtn">^</span></label>

		<div class="headTitle" id="Title" onclick="EPPZScrollTo.scrollVerticalToElementById('App', 50)">
			<span id="titLeft">Gene</span><span id="titRight">Targeter</span>
		</div>

		<div class="mainText">
			<p>
	            Creates a construct developed by the
	            <a href="http://web.mit.edu/nileslab/" target="_blank">Niles Lab at MIT</a>,
	            designed to deliver a 3' UTR post-transcriptional regulatory element
	            payload to a specific given gene in <a href="http://www.who.int/mediacentre/factsheets/fs094/en/" target="_blank"><i>Plasmodium falciparum</i></a>.
	        </p>
	        <p>
	            Future plans include custom plasmids, automatic gRNA selection, and
	            5'/3' UTR editing.
			</p>
		</div>

		<div id="App" class="app" onload="init()">
			<script src="//cdnjs.cloudflare.com/ajax/libs/socket.io/0.9.16/socket.io.min.js"></script>
			<!--FALLBACK SOCKET.IO LOCAL <script type="text/javascript" src="static/js/socket.io.min.js"></script> -->
			<script src="static/js/socketHandler.js"></script>
			<!--script src="static/js/app.js"></script-->
			<p>
				Upload a
				<a href="https://benchling.com" class="tooltip" target="_blank"
					tipTxt="A GenBank file (.gb) can
					be downloaded from Benchling clicking on the Information
					icon on the righthand tool menu and scrolling to the
					'Download' section.">
					GenBank file</a>
				annotated with the gene's sequence
				<a href="https://benchling.com" class="tooltip" target="_blank"
					tipTxt="Benchling displays 5'->3'
					annotations from left to right on the top strand. If your
					gene appears from right to left on the bottom strand, you
					can select the whole sequence -> right click -> Copy Special
					-> Reverse Complement -> paste in new file.">
					in 5'->3' sense.</a>
				</p>

				<p>
					For every gene uploaded, GeneTargeter outputs
					<ul>
						<li>
							<b>a GenBank file</b> containing the given gene annotated with
							the chosen left and right homologous regions, before chromosomal
							editing.
						</li>
						<li>
							<b>a GenBank file</b> containing the fully annotated plasmid
							vector designed to target the given gene.
						</li>
						<li>
							<b>a GenBank file</b> containing the edited chromosomal locus
							targeted by the construct.
						</li>
						<li><b>a CSV file</b> containing primers,
							gene fragment sequences and Klenow reaction oligos designed to
							assemble the new plasmid using
							<a href="https://dx.doi.org/10.1038/nmeth.1318" target="_blank">
								Gibson Assembly</a>.
							This file can be
							<a href="https://benchling.com" class="tooltip" target="_blank"
								tipTxt="Open the '+' drop-down menu in the top-right corner of
								Benchling's interface, select 'More', and select 'Import oligos'.
								Copy and paste the CSV file contents onto Benchling's field.">
								imported into Benchling</a>. Abbreviates primer names to fit on
								commercial tube labels with the <a href="" class="tooltip" tipTxt="The seven digit gene identifier
								code follows 'PF3D7_'.
								Oligo types include:
								LHR (Gibson overhang PCR primers)
								RHR (Gibson overhang PCR primers)
								gRNA (Klenow oligos for gRNA sequence)
								gBlock (gBlock sequencing primer)
								RecKlen (Klenow oligos for recoded region, if the region is small
								enough)
								Orientation refers to forward (F) and reverse (R) primers.">format:</a><br>

								Seven Digit Gene Identifier_Oligo type_Orientation
						</li>
						<li>
							<b>a text file</b> containing a message log and warnings
							issued during the <i>in silico</i> design process.
						</li>
					</ul>
				</p>
			<div class="small" id="controls">
				<p>
					Use homology regions and/or guide RNAs already annotated on files (if any):
					<input class="chkBox" type="checkbox" unchecked id="HRannChkBox">
				</p>
				<label class="btn" id="optionsBtn" onclick="toggleOptions()">Show more options</label>
				<div id=moreOptions class="expand">
					<div class="measuringWrapper">
						<br>
						<p>
							Guide RNAs will be selected automatically based
							on <a href="https://dx.doi.org/10.1038/nbt.3437" target="_blank">
							Doench et al. (2016)</a> on-target score and either
							<a href="https://dx.doi.org/10.1038/nbt.3437" target="_blank">
							Doench et al. (2016)</a> or
							<a href="http://crispr.mit.edu/about" target="_blank">
							Hsu et al. (2013)</a> off-target scores. Off-target scores
							are evaluated based on off-target gRNAs in the <i>P. falciparum</i>
							3D7 genome.
						</p>
						<p>
							Alternatively, you may
							manually annotate all
							<a href="https://benchling.com" class="tooltip" target="_blank"
								tipTxt="You can use Benchling's CRISPR tool, but you must
								create annotations on the sequence for all the gRNAs to be
								considered by GeneTargeter.">
								CRISPR guide RNAs</a>
							to be included in targeting. Guide RNA annotations should start
							with the word "gRNA" in the label, followed by
							<a href="https://benchling.com" class="tooltip" target="_blank"
								tipTxt="i.e., your best-ranked gRNA should be labeled
								'MyGene gRNA 1', your second best should be 'MyGene gRNA
								2, etc.'" class="tooltip">
								a number</a>
							indicating their order of preference. You may also annotate
							hand-picked left and right homologous regions on your file, if
							you do not want GeneTargeter to choose them automatically for you.
						</p><br>

						<div class="container">
							<div class="column-center">
								<p class="head">Right chromosomal integration homologous region</p>
								Size:<br>
								&nbsp&nbsp Minimum: &nbsp&nbsp<input class="txtInput" type="number" id="RHRMin" value="450"> bp <br>
								&nbsp&nbsp Preferred: <input class="txtInput" type="number" id="RHRPref" value="500"> bp <br>
								&nbsp&nbsp Maximum: &nbsp&nbsp<input class="txtInput" type="number" id="RHRMax" value="750"> bp <br>
								<p>
									Maximum distance from gene:
									<input class="txtInput" type="number" id="maxDistRHR" value="500">
									bp
								</p>
								<p>
									Search for best start/end point
									<input class="txtInput" type="number" id="optLowRHR" value="20">
									bp upstream and
									<input class="txtInput" type="number" id="optHighRHR" value="20">
									bp downstream of first valid start/end point
								</p>
								<p>
									Minimum melting temperature of the first and last
									<input class="txtInput" type="number" id="endsRHR" value="40"> bp
									of the right homologous region:
									<input class="txtInput" type="number" id="endsTempRHR" value="59"> ÂºC
								</p>
							</div>
							<div class="column-left">
								<p class="head">Left chromosomal integration homologous region</p>
								Size:<br>
								&nbsp&nbsp Minimum: &nbsp&nbsp<input class="txtInput" type="number" id="LHRMin" value="450"> bp <br>
								&nbsp&nbsp Preferred: <input class="txtInput" type="number" id="LHRPref" value="500"> bp <br>
								&nbsp&nbsp Maximum: &nbsp&nbsp<input class="txtInput" type="number" id="LHRMax" value="650"> bp <br>
								<p>
									Maximum distance from gRNA:
									<input class="txtInput" type="number" id="maxDistLHR" value="500"> bp
								</p>
								<p>
									Search for best start/end point
									<input class="txtInput" type="number" id="optLowLHR" value="20">
									bp upstream and
									<input class="txtInput" type="number" id="optHighLHR" value="10">
									bp downstream of first valid start/end point
								</p>
								<p>
									Minimum melting temperature of the first and last
									<input class="txtInput" type="number" id="endsLHR" value="40"> bp of
									the left homologous region:
									<input class="txtInput" type="number" id="endsTempLHR" value="55"> ÂºC
								</p>
							</div>
							<div class="column-right">
								<p class="head">Gibson Assembly</p>
								Homologous region size*:
								&nbsp&nbsp Minimum: &nbsp&nbsp<input class="txtInput" type="number" id="gibMin" value="30"> bp <br>
								&nbsp&nbsp Preferred: <input class="txtInput" type="number" id="gibPref" value="40"> bp <br>
								&nbsp&nbsp Maximum: &nbsp&nbsp<input class="txtInput" type="number" id="gibMax" value="50"> bp <br>
								<br>
								*Each primer is twice the homologous region size.
								<p>
									Minimum primer melting temperature:
									<input class="txtInput" type="number" id="gibTemp" value="65"> ÂºC
								</p>
								<p>
									Maximum melting temperature difference between primers:
									<input class="txtInput" type="number" id="gibTDif" value="5"> ÂºC
								</p>
							</div>
						</div>
						<div class="container">
							<div class="column-leftcenter">
								<p class="head">Guide RNA (gRNA) Selection</p>
								<p>
									Minimum GC content: &nbsp<input class="txtInput" type="number" id="gRNAGCContent" value="30"> % <br>
									Minimum <a href="https://doi.org/10.1038/nbt.3437" target="_blank">
										on-target score: </a>&nbsp<input class="txtInput" type="number" id="gRNAOnTargetCutoff" value="35"> % <br>
									Off-target <a href="http://crispr.mit.edu/about" class="tooltip" target="_blank"
										tipTxt="CFD scores are more accurate (Doench et al., 2016). Hsu et al. (2013) scores are used by crispr.mit.edu. Individual scores are aggregated as per crispr.mit.edu.">
										scoring method</a> for single hits:<br>
											<select id="gRNAOffTargetMethod" onchange="changeScoringMethod()">
												<option value="cfd" selected="selected">CFD (Doench et al., 2016)</option>
												<option value="hsu">Zhang Lab (Hsu et al. 2013)</option>
											</select> <br>
									Minimum off-target total score: </a>&nbsp<input class="txtInput" type="number" id="minOffTargetScore" value="0.35"> %<br>
									Maximum single hit off-target score:</a>&nbsp<input class="txtInput" type="number" id="maxOffTargetHitScore" value="30"> %<br>
								</p>
								<br><br>
							</div>
							<div class="column-right">
								<p class="head">
									Recoded Region
								</p>
								<p>
									Codon optimize to
									<select id="codonOptimizeOrg">
										<option value="T. gondii" selected="selected">T. gondii</option>
										<option value="P. falciparum 3D7">P. falciparum 3D7</option>
										<option value="P. vivax">P. vivax</option>
										<option value="E. coli K12">E. coli K12</option>
										<option value="S. cerevisiae">S. cerevisiae</option>
										<option value="H. sapiens">H. sapiens</option>
										<option value="R. norvegicus">R. norvegicus</option>
										<option value="scramble">Scramble</option>
									</select>
									<a href="http://omicsomics.blogspot.com/2009/04/is-codon-optimization-bunk.html" target="_blank">
										using</a>
									<select id="codonOptimStrat">
										<option value="CAI Maximization">CAI Maximization</option>
										<option value="Codon Sampling">Codon Sampling</option>
									</select>
								</p>
								<p>
									Minimum gBlock gene fragment size (will create Klenow oligos if smaller):
									<input class="txtInput" type="number" id="minFragSize" value="125">
									bp
								</p>
							</div>
						</div>
					</div>
				</div>
			</div>
			<p>
				<label class="btn" ><input type="file" id="geneFileForm" accept=".gb" multiple size="50" onchange="uploadFile()">
					<span> Upload .gb files </span></label>
			</p>

			<p>
				<label class="runBtn" id="run" onclick="askCredentials()">Select one or more GenBank gene files.</label>
			</p>

			<div class="fileList small" id="selectedFiles">
			</div>

			<p>
				<div class="fileList" id="outputLog"></div>
			</p>

			<div id="myModal" class="modal">

			  <!-- Modal content -->
			  <div class="modal-content" id="modContent">
			    <div class="modal-header">
			      <span class="close" onclick="closeModal()">Ã</span>
			      <h2>Hold up though...</h2>
			    </div>
			    <div class="modal-body">
					<p>
					  GeneTargeter uses DNA sequences that are currently not
					  available to the public. The full software will be
					  publicly available in the near future.
					</p>
					<p>
					  If you have the credentials to download the output, input
					  your passcode here:
					</p>
					<p>
						<input type="password" id="passcode" value="" onkeydown="checkKey()">
					</p>
					<p>
						<label class="runBtn" id="submitPassword" onclick="verifyCredentials()">Submit</label>
					</p>
			    </div>
			    <div class="modal-footer" id="modFooter">
			      <h3>Type passcode</h3>
			    </div>
			  </div>

			</div>


		</div>

		<div class="bottom">
			<a href="http://web.mit.edu/nileslab/" target="_blank">
				<p>
					<img src="static/assets/roundLogoNilesLab.png" width="400px">
				</p>
				<p>
					<img src="static/assets/WhiteLogoNilesLab.png" width="400px">
				</p>
			</a>
			<p>
	        Background Image:
	        <a href="https://commons.wikimedia.org/wiki/File:Plasmodium_falciparum_in_a_patient_presenting_with_pyrexia_and_headache.jpg" target="_blank">
	            <i>Plasmodium falciparum in a patient presenting with pyrexia and
	            headache.</i> O. Erhabor (2013)</a>
	        <br>
	        <a href="https://creativecommons.org/licenses/by-sa/3.0/deed.en" target="_blank">
	            CC BY-SA 3.0</a>
	    </p>
			<p>
					On-target gRNA scoring code and CFD scoring matrix obtained December 27, 2016 from <a href="https://dx.doi.org/10.1038/nbt.3437" target="_blank">Doench et al. (2016)</a> supplementary material.
			</p>
			<p>
					Off-target Zhang Lab scoring matrix obtained December 27, 2016 from the online <a href="http://crispr.mit.edu/about">MIT CRISPR tool</a> <a href="https://dx.doi.org/10.1038/nbt.2647">(Hsu et al., 2016)</a>.
			</p>
			<p>
				Off-target gRNA database built from <i>P. falciparum</i> genome first published by
				<a href="https://dx.doi.org/10.1038/nature01097" target="_blank"> Gardner et al. (2002)</a>
				and subsequently edited by the <a href="http://plasmodb.org/plasmo/" target="_blank">PlasmoDB</a> community.
			</p>
			<p>
					Codon frequency tables obtained July 17, 2016 from the <a href="http://www.kazusa.or.jp/codon/" target="_blank">Codon Usage Database</a> run by Kazusa DRI.
			</p>
			<br><br>
	    <p>
	        All other code: <a href="http://opensource.org/licenses/MIT" target="_blank">
					MIT License</a>, appropriately.
				<br>
				<a href="https://github.com/pablocarderam/GeneTargeter" target="_blank">
					GitHub src here.</a>
	            <br>
	            Developed by Pablo CÃ¡rdenas RamÃ­rez
	            <br>
	            <a href="mailto:pablocarderam@gmail.com?Subject=Hi%20there" target="_top" class="pcr">
					pablocarderam@gmail.com</a>
			</p>
		</div>
	</body>
</html>
